roachpb/batch.go

```diff
@@ -176,11 +176,264 @@ func (ba *BatchRequest) Methods() []Method {
 }
 
 // CreateReply creates replies for each of the contained requests, wrapped in a
-// BatchResponse.
+// BatchResponse. The response objects are batch allocated to minimize
+// allocation overhead which adds a bit of complexity to this function.
 func (ba *BatchRequest) CreateReply() *BatchResponse {
 	br := &BatchResponse{}
+	br.Responses = make([]ResponseUnion, len(ba.Requests))
+
+	var counts struct {
```

Comment:
I considered this while writing this PR, but notice that if you add a new request/response to `Batch{Request,Response}` and don't populate it here you'll get a panic when trying to implement it. I discovered this when I accidentally left out `Increment{Request,Response}`. Another easy error here which is caught is incrementing the wrong count or using the wrong count when constructing the buffer. But as long as all of the requests are tested by other code, we'll hit panics when trying to populate the responses.

Push-back if you see a type of error here that wouldn't be caught be existing tests.